plugin/path_rotate_root_creds.go

```diff
@@ -33,22 +36,60 @@ func (b *backend) pathRotateCredentialsUpdate(ctx context.Context, req *logical.
 	if err != nil {
 		return nil, err
 	}
+	oldPassword := engineConf.ADConf.BindPassword
 
+	// Update the password remotely.
 	if err := b.client.UpdateRootPassword(engineConf.ADConf, engineConf.ADConf.BindDN, newPassword); err != nil {
 		return nil, err
 	}
-
 	engineConf.ADConf.BindPassword = newPassword
+
+	// Update the password locally.
+	if pwdStoringErr := storePassword(ctx, req, engineConf); pwdStoringErr != nil {
+		// We were unable to store the new password locally. We can't continue in this state because we won't be able
+		// to roll any passwords, including our own to get back into a state of working. So, we need to roll back to
+		// the last password we successfully got into storage.
+		if rollbackErr := b.rollBackPassword(ctx, engineConf, oldPassword); rollbackErr != nil {
+			return nil, fmt.Errorf("unable to store new password due to %s and unable to return to previous password due to %s, configure a new binddn and bindpassword to restore active directory support", pwdStoringErr, rollbackErr)
+		}
+		return nil, fmt.Errorf("unable to update password due to storage err: %s", pwdStoringErr)
+	}
+
+	// Respond with a 204.
+	return nil, nil
+}
+
+// rollBackPassword uses naive exponential backoff to retry updating to an old password,
+// because Active Directory may still be propagating the previous password change.
+func (b *backend) rollBackPassword(ctx context.Context, engineConf *configuration, oldPassword string) error {
+	var err error
+	for i := 0; i < 10; i++ {
+		waitSeconds := math.Pow(float64(i), 2)
```

Comment:
Did you instead mean: `math.Pow(2, float64(i))` which is often seen?  But tbh I don't think it matters much. The way it currently is is just a more gradual increase.

Reply:
Nope, I wanted it the first way. I'm trying to square i, basically, just a totally simple increase of the seconds. Comes out just how I want: https://play.golang.org/p/EUDzgjn9v-D